File: 0029-divide-two-integers/0029-divide-two-integers.java
class Solution {
    public int divide(int dividend, int divisor) {
        if (dividend == divisor) {
            return 1;
        }
        
        boolean isNeg = false;
        
        if ((dividend < 0 && divisor > 0) || (dividend > 0 && divisor < 0)) {
            isNeg = true;
        }
        
        long n = Math.abs((long)dividend), d = Math.abs((long)divisor);
        long res = 0;
        
        while (n >= d) {
            int cnt = 0;
            while (n >= (d << (cnt + 1))) {
                cnt++;
            }
            res += (1 << cnt);
            n = n - (d << (cnt));
        }
        
        if (res == (1 << 31)) {
            return isNeg ? Integer.MIN_VALUE : Integer.MAX_VALUE;
        }
        
        return isNeg ? (int)-res : (int)res;
    }
}

File: 0029-divide-two-integers/0029-divide-two-integers.cpp
class Solution {
public:
    typedef long long ll;
    
    int divide(int dividend, int divisor) {
        int sign = 1;
        
        if ((dividend >= 0 && divisor < 0) || (dividend < 0 && divisor > 0)) {
            sign = -1;
        }
        
        ll num = abs(ll(dividend));
        ll den = abs(ll(divisor));
        // cout << "num is " << num << " and den is " << den << "\n";
        ll res = 0;
        
        while (num >= den) {
            int power = 0;
            while (num >= (den << power)) {
                power++;
            }
            power--;
            num = num - (den << power);
            res = res + (1 << power);
        }
        
        if (res == (1 << 31)) {
            return sign == 1 ? INT_MAX : INT_MIN;
        }
        
        return sign * res;
    }
};